gateway/ln-gateway/src/lib.rs

```diff
@@ -84,6 +93,221 @@ pub const DEFAULT_FEES: RoutingFees = RoutingFees {
+
+    /// Gateway webserver listen address
+    #[arg(long = "listen", env = "FM_GATEWAY_LISTEN_ADDR")]
+    pub listen: SocketAddr,
+
+    /// Public URL from which the webserver API is reachable
+    #[arg(long = "api-addr", env = "FM_GATEWAY_API_ADDR")]
+    pub api_addr: Url,
+
+    /// Gateway webserver authentication password
+    #[arg(long = "password", env = "FM_GATEWAY_PASSWORD")]
+    pub password: String,
+
+    /// Configured gateway routing fees
+    /// Format: <base_msat>,<proportional_millionths>
+    #[arg(long = "fees", env = "FM_GATEWAY_FEES")]
+    pub fees: Option<GatewayFee>,
+}
+
+pub struct Gatewayd {
+    registry: ClientModuleGenRegistry,
+    lightning_mode: LightningMode,
+    data_dir: PathBuf,
+    listen: SocketAddr,
+    api_addr: Url,
+    password: String,
+    fees: Option<GatewayFee>,
+}
+
+impl Gatewayd {
+    pub fn new() -> anyhow::Result<Gatewayd> {
+        let mut args = std::env::args();
+
+        if let Some(ref arg) = args.nth(1) {
+            if arg.as_str() == "version-hash" {
+                println!("{}", env!("FEDIMINT_BUILD_CODE_VERSION"));
+                std::process::exit(0);
+            }
+        }
+
+        // Read configurations
+        let GatewayOpts {
+            mode,
+            data_dir,
+            listen,
+            api_addr,
+            password,
+            fees,
+        } = GatewayOpts::parse();
+
+        info!(
+            "Starting gatewayd (version: {})",
+            env!("FEDIMINT_BUILD_CODE_VERSION")
+        );
+
+        Ok(Self {
+            registry: ClientModuleGenRegistry::new(),
+            lightning_mode: mode,
+            data_dir,
+            listen,
+            api_addr,
+            password,
+            fees,
+        })
+    }
+
+    pub fn with_module<T>(mut self, gen: T) -> Self
+    where
+        T: ClientModuleGen + 'static + Send + Sync,
+    {
+        self.registry.attach(gen);
+        self
+    }
+
+    pub fn with_default_modules(self) -> Self {
+        // Gateway module will be attached when the federation clients are created.
+        self.with_module(MintClientGen)
+            .with_module(WalletClientGen::default())
+    }
+
+    pub async fn run(self) -> anyhow::Result<()> {
+        TracingSetup::default().init()?;
+
+        let decoders = self
+            .registry
+            .decoders(DEFAULT_MODULE_KINDS.iter().cloned())?;
+
+        let client_builder = StandardGatewayClientBuilder::new(
+            self.data_dir.clone(),
+            self.registry.clone(),
+            LEGACY_HARDCODED_INSTANCE_ID_MINT,
+        );
+
+        let db = Database::new(
+            fedimint_rocksdb::RocksDb::open(self.data_dir.join(DB_FILE))?,
+            decoders.clone(),
+        );
+
+        let mut tg = TaskGroup::new();
+        let rx = self.start_gateway(&mut tg, client_builder, db).await?;
+        rx.await?;
+        info!("Gatewayd exiting...");
+        Ok(())
+    }
+
+    pub async fn start_gateway(
```

Comment:
Shouldn't this be private since without it the modules defined in `Gatewayd` will be discarded, which is counterintuitive?

Reply:
Ah yes, this was overlooked when copy/pasted. It should be private